src/blockdata/locktime.rs

```diff
@@ -0,0 +1,631 @@
+// Rust Bitcoin Library
+// Written in 2022 by
+//     Tobin C. Harding <me@tobin.cc>
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Provides type [`LockTime`] that implements the logic around nLockTime/OP_CHECKLOCKTIMEVERIFY.
+//!
+//! There are two types of lock time: lock-by-blockheight and lock-by-blocktime, distinguished by
+//! whether `LockTime < LOCKTIME_THRESHOLD`.
+//!
+
+use core::{mem, fmt};
+use core::cmp::{PartialOrd, Ordering};
+use core::convert::TryFrom;
+use core::str::FromStr;
+use core::num::ParseIntError;
+
+use crate::consensus::encode::{self, Decodable, Encodable};
+use crate::io;
+use crate::prelude::*;
+
+/// The Threshold for deciding whether a lock time value is a height or a time (see [Bitcoin Core]).
+///
+/// Bitcoin is able to safely use this value because a block height greater than 500,000,000 would
+/// never occur because it would represent a height in approximately 9500 years. Conversely, block
+/// times under 500,000,000 will never happen because they would represent times before 1986 which
+/// are, for obvious reasons, not useful within the Bitcoin network.
+///
+/// [Bitcoin Core]: https://github.com/bitcoin/bitcoin/blob/9ccaee1d5e2e4b79b0a7c29aadb41b97e4741332/src/script/script.h#L39
+const LOCK_TIME_THRESHOLD: u32 = 500_000_000;
+
+/// A transaction with nLocktime==0 is able to be included immediately in any block.
+pub const ZERO: LockTime = LockTime::Blocks(Height(0)); // This could just as well have been LockTime::Time(Seconds(500_000_000))
```

Comment:
I'm not convinced about this. Its only useful as a shortcut to `LockTime::from_consensus(0)`, right? Its not immediately apparent without domain knowledge what zero means or, when reading the comment, why there is not a `LockTime::Seconds(Height(LOCK_TIME_THRESHOULD))` constant.

Reply:
>  Its not immediately apparent without domain knowledge what zero means or

That's a good point.

but...

> `LockTime::from_consensus(0)` 

doesn't change it all. One still doesn't know what this zero means. Possibly they 


Is it common to rely on value `0` anywhere? We should probably get rid of it, and IIRC miniscript is the only place that treats it magically?

If we do need it, maybe `pub const NO_TIMELOCK` would be better (with the same comment). How exactly it is initialized is secondary, IMO. The result is the same. :)

Reply:
Maybe Have another variant `Inactive` instead? Parse `0` as `Inactive` and reject `0` in `Height` constructor?

Reply:
Please no more new names :(

Zero is useful all the time because it's the most common fixed locktime value, and indicates that the user does not want any locktime on the transaction. During verification it can also be used as "assumed valid" locktime.

Reply:
Just to confirm: `0` has no special treatment in the consensus, right? It is just a convention and relies on `0` being greater or equal to any non-signed integer.

 I guess `ZERO` constant is OK. We should probably expand the documentation to explain that it's a convention, default and some software might treat it specially, but all in all it is just `Block(0)`.

Reply:
Maybe we should really leave these functions out for now, and let the user do whatever they want/mean. :D

```rust
match lock_time {
    LockTime::Height(height) => height >= get_height(),
    LockTime::Time(time) => time >= get_time(),
}
```

is not that bad. :D

Reply:
I've removed all the `is_satisifed*` methods _except_ `pub fn is_satisfied_by(&self, height: Height, time: Time) -> bool`. At least this function serves as a reference for which way the `>=` signs go if folks want to use the code above.

Reply:
- `is_satisfied` with single parameter removed.
- ZERO left as is
- Add example and attempted to explain in words also (see rustdocs on `is_satisfied_by`